keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+>   "default": first default value of  `x_i` or fail
+>}
+>```
+
+The skeleton schema especially lacks:
+* all value validations like `pattern`, `format`, `minValue`, `maxValue`, ...
+* all propositional operators like `anyOf`, `allOf`, `oneOf`, `not`.
+
+The skeleton schema `skel(s)` puts
+* less or equal constraints on `type` than `s`.
+* every field specified by a `properties` key in `s` is specified in `skel(s)`.
+* if there are no default values under `allOf`, `anyOf`, `oneOf`, `not` in `s`, for each JSON path the default matches in `s` and `skel(s)`.
+
+The computation of `skel(s)` is `O(size(s))` and `size(skel(s)) <= size(s)`.
+
+Note that a field might be constrained by the `type` construct in the full OpenAPI validation schema, but not in its skeleton. This is fine because we have to support polymorphic fields like `IntOrString`, but avoid  `allOf`, `anyOf`, `oneOf`, `not` in the skeleton.
+
+**Property:** if the OpenAPI validation schema applies to an object, so does its skeleton schema.
+
+Pruning and defaulting will be implemented based on the skeleton schema of the specified OpenAPI validation schema.
+
+Optionally for debugging, the skeleton schema derived from the validation schema by the apiextensions-apiserver can be stored in the CRD status.
+
+### Example 8
+
+>Assume 
+>
+>```
+>s := {
+>    "anyOf": [
+>        {"properties": {"a": {"type": "string"}}},
+>        {"properties": {"a": {"type": "integer"}, "b": {"type": "string"}}}
+>    ],
+>    "properties": {"c": {}}
+>}
+>```
+>
+>Then 
+>
+>```
+>skel(s) = {
+>    "properties": {
+>	    "a": {},
+>	    "b": {"type": "string"},
+>	    "c": {}
+>    }
+>}
+>```
+>
+> The `type` of `a` does not match on all paths. Hence, it is omitted in the skeleton. In contrast, `b` has a unique `type` of `"string"`, so it stays in the skeleton.
+
+### Types and Formats
+
+Note, that having less `type` constraints in the skeleton than in the whole OpenAPI validation schema means that admission and conversion will see possibly wrong types. The final validation in the registry validation phase will check for the complete OpenAPI validation schema and catch those type errors.
+
+**Question:** we could extent the `skel` function to keep a list of `type` values. As long as all branches define the type, we can add `anyOf: [{type: "type1", type: "type2, ….}]` to the skeleton. If one branch does not define the type though, this is still not possible.
+
+In native types, we have custom unmarshallers for date / timestamps. In OpenAPI these would be rendered as `{type: "string", format: "date"}`. With the proposed skeleton algorithm, we would not verify these fields before full OpenAPI validation in the registry. We could move non-contradicting `format` constraints into the skeleton as well, like we do for `type` already. Then admission would be protected from invalid format (if admission uses Golang decoding, this might be relevant to get proper error messages).
```

Comment:
Is this important, beyond trying to mimic what we did for core types, which grew organically?  I talk to users of admission webhooks and I haven't heard anyone complain that they are seeing insufficiently validated data.

Reply:
If we don't do type validation during decoding, the admission webhooks have to cope with wrong types **and report good error messages**. Technically this is not a problem, but it puts more responsibility on the admission webhook implementation. If we do that already in decoding, the behaviour might be more uniform.